src/casp/services/api/clients/matching/matching_client.py

```diff
@@ -0,0 +1,188 @@
+from typing import List, NamedTuple
+
+from google.cloud import aiplatform_v1
+from google.cloud.aiplatform.matching_engine.matching_engine_index_endpoint import MatchNeighbor
+
+from casp.services.api import clients
+from casp.services.db import models
+
+"""
+Contains a common interface for querying the matching engine regarless of API (e.g. gRPC vs REST)
+"""
+
+
+class MatchResult(NamedTuple):
+
+    class Neighbor(NamedTuple):
+        """
+        Neighbor represents a neighbor of a feature vector in the matching service.
+        """
+
+        id: str
+        distance: float
+        feature_vector: List[float] = None
+
+    class NearestNeighbors(NamedTuple):
+        """
+        NearestNeighbors represents the result of a nearest neighbors request to the matching service. There should be 1
+        per query that was sent.
+        """
+
+        neighbors: List["MatchResult.Neighbor"] = []
+
+    """
+    MatchResult represents the result of a match request to the matching service.
+    """
+    matches: List["MatchResult.NearestNeighbors"] = []
+
+    def concat(self, other: "MatchResult") -> "MatchResult":
+        """
+        Concatenate the results of two match requests.
+
+        :param other: The other match result to concatenate.
+
+        :return: The concatenated match result as a new instance of MatchResult.
+        """
+        return MatchResult(matches=self.matches + other.matches)
+
+
+class MatchingClient:
+    def __init__(self, index: models.CASMatchingEngineIndex) -> None:
+        self.index = index
+
+    async def match(self, queries: List[List[float]]) -> MatchResult:
+        """
+        Match queries against the specified index.
+
+        :param index: The index to match against.
+        :param queries: The queries to match.
+
+        :return: The match result.
+        """
+        pass
+
+    @staticmethod
+    def from_index(index: models.CASMatchingEngineIndex) -> "MatchingClient":
+        """
+        Create a new MatchingClient from the specified index.
+
+        :param index: The index to create the client from.
+
+        :return: The new client.
+        """
+        if index.is_grpc:
+            return MatchingClientGRPC(index=index)
+        else:
+            return MatchingClientREST(index=index)
+
+
+class MatchingClientGRPC(MatchingClient):
```

Comment:
This and MatchingClientREST could maybe each benefit from a short comment explaining how they differ and why they both exist, as opposed to just having the one client.

Reply:
Totally.  Will do